databricks_cli/tunnel/api.py

```diff
@@ -0,0 +1,236 @@
+# Databricks CLI
+# Copyright 2021 Databricks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"), except
+# that the use of services to which certain application programming
+# interfaces (each, an "API") connect requires that the user first obtain
+# a license for the use of the APIs from Databricks, Inc. ("Databricks"),
+# by creating an account at www.databricks.com and agreeing to either (a)
+# the Community Edition Terms of Service, (b) the Databricks Terms of
+# Service, or (c) another written agreement between Licensee and Databricks
+# for the use of the APIs.
+#
+# You may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+from pathlib import Path
+
+import requests
+import tornado
+from cryptography.hazmat.primitives.asymmetric import rsa
+from tornado.ioloop import IOLoop
+
+from databricks_cli.clusters.api import ClusterApi
+from databricks_cli.commands.api import CommandApi, ExecutionContextApi
+from databricks_cli.tunnel.server import TunnelConfig, StreamingServer
+
+
+def generate_key_pair():
+    # TODO(tunneling-cli): support customer-supplied key / passphrase encryption
+    from cryptography.hazmat.primitives import serialization
+
+    key = rsa.generate_private_key(
+        public_exponent=65537,
+        key_size=4096
+    )
+
+    private_key = key.private_bytes(
+        encoding=serialization.Encoding.PEM,
+        format=serialization.PrivateFormat.OpenSSH,
+        encryption_algorithm=serialization.NoEncryption()
+    )
+
+    public_key = key.public_key().public_bytes(
+        encoding=serialization.Encoding.OpenSSH,
+        format=serialization.PublicFormat.OpenSSH
+    )
+
+    return private_key, public_key
+
+
+class TunnelApi:
+    def __init__(self, api_client, debug=False):
+        self.cluster_client = ClusterApi(api_client)
+        self.command_client = CommandApi(api_client)
+        self.exec_ctx_client = ExecutionContextApi(api_client)
+        self._api_client = api_client
+        self._default_local_ssh_dir = "~/.ssh"
```

Comment:
Just curious, this might not support Windows?  Is that OK?

Reply:
good point. cc: @jaipreet-s 

I think Windows users might be able to use WSL / OpenSSH? Or in the worst case, they might have to use docker container running a unix-based system...

Currently, in the context of the tunneling cli, what determines whether windows users can use it:
1. Can these keys be generated on windows?
2. Can the ssh command on Windows find the default ssh key location?

Related to 1, we're currently using Python's cryptography lib and std `Pathlib`. As long as cryptography lib is compatible with Windows and Python's `Pathlib` + path that we're using is abstract enough and compatible s.t. it works with Windows filesystem, then this req is satisfied. [This](https://stackoverflow.com/questions/4028904/what-is-the-correct-cross-platform-way-to-get-the-home-directory-in-python) and [this](https://stackoverflow.com/questions/2953834/windows-path-in-python) suggests that our implementation might work cross-platform.

2 needs some investigation. Although [this doc](https://docs.microsoft.com/en-us/windows-server/administration/openssh/openssh_keymanagement) suggests it might be possible.